### DISCUSSION TITLE
Slower startup on M1 mac?

### DISCUSSION THREAD

**Discussion by ptonner** [kkinnear/zprint#302] [Q&A]
Is it typical to have a slower startup time on M1 mac? I've tried `zprint` on two M1 machines and they both had a startup time of around 300 ms. Unfortunately this is noticeable enough to cause a slight lag when trying to run `zprint` on every file save. Just wondering if this is expected?

**Comment by kkinnear** [comment]
I happen to develop on a M1 Mac, so it is easy for me to try it.  When I run `zprintm-1.2.7` the first time it runs it generally takes a bit over 300ms, and when I run it again it consistently runs in just under 230ms.  Which is interesting, in that it starts up much more quickly on an Intel Mac.  Presumably the Rosetta support on macOS is adding some startup time.  Originally I was just pleased that it ran at all...

I have noticed that graalVM now seems to run on M1 chips, so I'll look into building an executable specifically for macOS M1 distributions.  I usually find upgrading to a new graalVM version to be ... problematic, so this is unlikely to be a simple "rebuild" and have it all work.  Though that has sometimes happened, it doesn't seem to be the most common experience.

Hopefully when I release 1.2.8, I can have three executables instead of just two, with a new one for macOS M1 support, which starts up faster and hopefully runs faster as well.  We'll see.

**Comment by kkinnear** [comment]
Well, I seem to have broken the code to use a later version of graalVM that supports M1 Macs.  I have gotten a zprint build to work, at least apparently. At least to do some simple things -- I don't know if it will hold up with concerted testing. 

That said, the graalVM M1 build takes about 23 ms to start up.  Which is pretty nice.  More interestingly, I have a very big file that takes 4.7 secs to format with the regular zprintm-1.2.7, (which is the Intel version running on my M1 Mac), and with zprintm1-1.2.8, (the new M1 graalVM build) it takes only 1.3 seconds.  Which is an amazing speed up!  

So, if I can ensure that the M1 version passes my tests, it looks like I will be distributing three pre-built binaries when I release 1.2.8.

Thanks for bringing this up, and causing me to look into the M1 graalVM!

**Comment by ptonner** [comment]
awesome, thank you for looking into it! I only noticed the difference when switching over to a linux machine and getting the blazing fast sub-30ms startup times and wondering why I wasn't getting the same on my M1 machine. Super excited for the next release!

**Comment by kkinnear** [comment]
Finally released `1.2.8` with a new pre-compiled binary: `zprintma-1.2.8` for Apple Silicon.  I hope that it works well for you.  Thank you again for bringing this up!
